python/rikai/spark/functions/vision.py

```diff
@@ -234,3 +237,79 @@ def spectrogram_image(
     return Image.from_array(
         np.frombuffer(output, np.uint8).reshape([size, size, 3]), output_uri
     )
+
+
+@udf(returnType=ArrayType(SegmentType()))
+def video_segmentor(
+    video: Union[VideoStream, YouTubeVideo],
+    threshold: float = 0.3,
+    fps: int = 30,
+) -> list:
+    """Returns list of video Segments from VideoStream or YouTubeVideo by
+    thresholding sum-of-absolute differences.
```

Comment:
Are there popular alternatives to sum-of-absolute differences? If so, how might we make this configurable so we can use other methods?

Reply:
[ffmpeg's](http://www.ffmpeg.org/ffmpeg-filters.html#select_002c-aselect) scene detection uses simple threshold based detection without much flexibility.

[pyscenedetect](https://pyscenedetect.readthedocs.io/en/latest/reference/detection-methods/) offers a similar threshold based scene detector in addition to a more sensitive content-aware detector. We could use their base `SceneDetector` to make custom ones.

Reply:
do you think that ffmpeg overall is just too low-level of a tool for us to use directly in rikai? feels like in at least two things, OpenCV and pyscenedetect there are higher level APIs that are available?

Reply:
low-level like a fox! FFmpeg can help to decode all kinds of streams from multimedia containers.

I also found that ffmpeg can be built to run [opencv as filters](http://underpop.online.fr/f/ffmpeg/help/ocv.htm.gz).

Reply:
Finding relevant tensorflow-io pulls on this topic:

[#11 ](https://github.com/tensorflow/io/issues/11)
[#144](https://github.com/tensorflow/io/issues/144)
[#30](https://github.com/tensorflow/io/pull/30)

Reply:
maybe something like this [example](https://github.com/kkroening/ffmpeg-python/blob/master/examples/tensorflow_stream.py) that produces images from bytes without converting to numpy.

Reply:
[skvideo scene detection](http://www.scikit-video.org/stable/examples/scene.html#sceneexample)